
<h1 style="text-align: center;">CIÊNCIA DE DADOS</h1>
<h1 style="text-align: center;">Roteiro de Atividade Prática</h1>
<br>
<br>

Nome: ______________________________________________________________________________________      

Turma: ______________


**Componente:** Aprendizagem de Máquina
<br>
**Unidade Curricular:** Modelos, Algoritmos e Estimadores
<br>
**Tema da Semana:** Regressão Logística (Parte 2)
<br>


# Aula 4: Exercícios de fixação

## Tarefa
- Execute o código abaixo.

- Se as bibliotecas não tiverem instaladas, instale-as.

- Acompanhe as instruções.

- Preencha o código onde está #TO DO

- Responda as perguntas e conclua a atividade


In [ ]:
### Código

### Imports de dependências

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import accuracy_score, precision_score, recall_score, make_scorer, f1_score
from sklearn.model_selection import cross_val_score
import warnings
from sklearn.exceptions import ConvergenceWarning

### Configurações

In [ ]:
SEED = 42
np.random.seed(SEED)
warnings.filterwarnings("ignore", category=ConvergenceWarning)
warnings.filterwarnings("ignore")

In [ ]:
cols = ['Aroma', 'Flavor', 'Aftertaste', 'Acidity', 'Body',
       'Balance', 'Uniformity', 'Clean.Cup', 'Sweetness',
       'Moisture']

### Leitura dos dados

In [ ]:
df = pd.read_csv('arabica_data_cleaned.csv', sep=',')

df = df.dropna(axis=0, subset=cols + ['Total.Cup.Points'])

X = df[cols]

# Condições e valores para cada categoria
conditions = [
    df['Total.Cup.Points'] >= 85,                # Condição para "high"
    (df['Total.Cup.Points'] >= 79) & (df['Total.Cup.Points'] < 85), # Condição para "medium"
    df['Total.Cup.Points'] < 79                  # Condição para "low"
]
values = [2, 1, 0]

# Cria uma nova coluna com a classificação
df['quality'] = np.select(conditions, values)

y = df['quality']

# Normalizar as características usando StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


In [ ]:
df['quality']

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=SEED)

### Criação do modelo

In [ ]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=1000)
model.fit(X_train, y_train)


LogisticRegression(max_iter=1000, multi_class='multinomial')

In [ ]:
y_pred = model.predict(X_test)

# Definição das métricas precisão, revocação, f1 score e acurácia

In [ ]:
precision_scorer = make_scorer(precision_score, average='weighted')

recall_scorer = make_scorer(recall_score, average='weighted')

f1_scorer = make_scorer(f1_score, average='weighted')

accuracy_scorer = 'accuracy'

# Avaliação do modelo

In [ ]:
accuracy_scores = cross_val_score(model, X, y, cv=5, scoring=accuracy_scorer)
print(f'Cross-Validation Accuracy Scores: {accuracy_scores}')
print(f'Mean Accuracy Score: {accuracy_scores.mean():.4f}')

precision_scores = cross_val_score(model, X, y, cv=5, scoring=precision_scorer)
print(f'Cross-Validation Precision Scores: {precision_scores}')
print(f'Mean Precision Score: {precision_scores.mean():.4f}')

recall_scores = cross_val_score(model, X, y, cv=5, scoring=recall_scorer)
print(f'Cross-Validation Recall Scores: {recall_scores}')
print(f'Mean Recall Score: {recall_scores.mean():.4f}')

f1_scores = cross_val_score(model, X, y, cv=5, scoring=f1_scorer)
print(f'Cross-Validation F1 Scores: {f1_scores}')
print(f'Mean F1 Score: {f1_scores.mean():.4f}')

Cross-Validation Accuracy Scores: [0.84790875 0.99618321 1.         0.94656489 0.79389313]
Mean Accuracy Score: 0.9169
Cross-Validation Precision Scores: [0.88846052 0.99620048 1.         0.94976186 0.86746948]
Mean Precision Score: 0.9404
Cross-Validation Recall Scores: [0.84790875 0.99618321 1.         0.94656489 0.79389313]
Mean Recall Score: 0.9169
Cross-Validation F1 Scores: [0.83745842 0.99614099 1.         0.93403    0.78407572]
Mean F1 Score: 0.9103


### Execute o código para efetuar a avaliação de desempenho do modelo e responda.

## Tarefa
Execute o código para efetuar a avaliação de desempenho do modelo utilizando validação cruzada e reponda.
- Qual a precisão da avaliação utilizando Cross Validation? O que isto significa?
- Qual a revocação na avaliação utilizando Cross Validation? O que isto significa?
- Qual o f1 score na avaliação utilizando Cross Validation? O que isto significa?
- Qual a acurácia na avaliação utilizando Cross Validation? O que isto significa?














• Qual a precisão? Esse resultado pode ser considerado bom?

• Resposta:
• Precisão obtida: (Substitua pelo valor calculado, por exemplo, 0,94);
• Análise: A precisão de 0,94 indica que o modelo classificou corretamente 94% dos exemplos positivos entre as previsões feitas. Esse valor sugere que o modelo
tem uma taxa razoável de erros ao classificar exemplos de outras classes, mas ainda está em um nível aceitável para muitos contextos de aplicação.

• Qual o recall? Esse resultado pode ser considerado bom?

• Resposta:
• Recall obtido: (Substitua pelo valor calculado, por exemplo, 0,91);
• Análise: Com um recall de 0,91, o modelo conseguiu identificar 91% dos exemplos da classe positiva, indicando que ainda existem alguns exemplos da classe
positiva que ele não conseguiu capturar. Isso pode ser aceitável dependendo do problema, mas há espaço para melhorias se for essencial minimizar os falsos
negativos.

• Qual o F1-Score? Esse resultado pode ser considerado bom?

• Resposta:
• F1-Score obtido: (Substitua pelo valor calculado, por exemplo, 0,91);
• Análise: O F1-Score de 0,91 representa um equilíbrio entre precisão e recall. Embora não esteja próximo de 1, um F1-Score de 0,91 é considerado bom e indica que
o modelo mantém um equilíbrio adequado entre evitar falsos positivos e capturar verdadeiros positivos.

• Qual a acurácia? Esse resultado pode ser considerado bom?

• Resposta:
• Acurácia obtida: (Substitua pelo valor calculado, por exemplo, 0,91);
• Análise: A acurácia de 0,91 significa que o modelo classificou corretamente 91% de todas as amostras, o que é bom, mas não excelente. Em um problema de
classificação multiclasse, esse valor é razoável e indica que o modelo tem um desempenho confiável.